### DISCUSSION TITLE
Jest Global Setup/Teardown in NX Workspace

### DISCUSSION THREAD

**Discussion by wolframm** [nrwl/nx#13316] [General]
Hi! I cannot figure out how to specify a Jest 29 global setup/teardown function for a NX 15 typescript workspace. 

I want to run global setup (starting various servers etc..) once before any test targets. There is a brief mention of global setup/teardown in the docs (https://nx.dev/packages/jest#global-setup/teardown-with-nx-libraries). However, I have no idea where I should place the global setup file and where/how to instruct Jest to find it and run it whenever tests are run in in any library. 

I would really appreciate a brief example, showing where the global setup file must be placed and how to point Jest to it (so it always runs when testing any typescript library in my workspace).

I posted this message also on Slack: https://nrwlcommunity.slack.com/archives/CMFKWPU6Q/p1669122586239249

**Comment by wolframm** [comment]
The solution was kindly provided on Slack. Please close/delete this discussion.

**Reply by twittwer** [reply]
Because Slack messages are preserved for 90 days only, we should keep the answer here.
Thx to @leon for providing the following answer:

> In you jest.preset.js in the root you can add the globalSetup script
```
const nxPreset = require('@nrwl/jest/preset').default
const path = require('path')

/** @type {import('@jest/types').Config.InitialOptions} */
module.exports = {
  ...nxPreset,
  globalSetup: path.join(__dirname, 'jest-global-setup.js'),
```

> Then in jest-global-setup.js
```
module.exports = async function globalSetup() {
```

> this will run once at startup

**Comment by wolframm** [comment]
The above solution by [leon](https://github.com/leon) works for running code inside 'jest-global-setup.js' or any file it references using relative or absolute import paths. Unfortunately it alone is not able to resolve library module imports (@myorg/my-setup-lib). I made a detailed post here: https://github.com/nrwl/nx/issues/13345#issue-1461230549

It would be great if someone had a solution for it.
